modyn/supervisor/internal/triggers/cost/incorporation_latency_tracker.py

```diff
@@ -0,0 +1,80 @@
+        curve."""
+
+    @property
+    def cumulative_latency_regret(self) -> float:
+        return self._cumulative_latency_regret
+
+    def add_latency(self, regret: float, period_duration: float) -> float:
+        """Add the new regret from the last interval to the cumulative regret.
+
+        Applicable if the regret can uniformly be aggregated to a scalar for every reporting data
+        batch. (e.g. batch drift distance)
+
+        This assumes uniformity in the aggregated regret value. If the regret is not uniform.
+        E.g. when it's build from the per-sample regrets of samples with varying timestamps, we
+        cannot simply assume that every regret component can be weighted by the period duration when adding it to
+        the cumulative regret. Some samples might only have arrived at the end of the period and thus
+        need a smaller weight. Consider using `add_latencies`.
+
+        Args:
+            regret_sum: The new regret value at the end of the interval.
+            interval_duration: The duration of the last period in seconds
+
+        Returns:
+            Most recent cumulative regret value.
+        """
+        self._current_regret += regret
+        self._cumulative_latency_regret += self._current_regret * period_duration
+
+        return self._cumulative_latency_regret
+
+    def add_latencies(
+        self,
+        regrets: list[tuple[int, float]],
+        start_timestamp: int,
+        period_duration: float,
+    ) -> float:
+        """Add the new regret after computing the regret sum from a list of
+        per-sample regrets.
+
+        Addressed the non-uniformity in the regret values by computing the regret sum from the per-sample.
+
+        Args:
+            regrets: List of regrets for each sample in the last period with their timestamps.
+            start_timestamp: The timestamp of the start of the last period.
+            end_timestamp: The timestamp of the end of the last period.
+
+        Returns:
+            Most recent cumulative regret value.
+        """
+        # We split the regrets into two parts: those that arrived in the last period and those that have been
+        # around for longer. In the latency cumulation (area under curve), the first make up a triangular shape,
+        # while the second contribute to the rectangle.
+
+        end_timestamp = start_timestamp + period_duration
+        regrets_durations = [(end_timestamp - timestamp, regret) for timestamp, regret in regrets]
+        new_regret = sum(regret for _, regret in regrets)
+
+        # rectangular shape of the area under the curve for the recently arrived regrets
+        new_regret_latency = sum(duration * regret for duration, regret in regrets_durations)
+
+        # old regret units that still contribute to area under curve (rectangular shape)
+        old_regret_latency = self._current_regret * period_duration
+
+        self._current_regret += new_regret
```

Comment:
I am trying to understand the meaning of the two local variables but I don't understand them yet.

My understanding is the following: If we have 10 samples since the last trigger, and now see a batch that has a duration of 20, we would want to calculate the regret of the batch and add 10 * 20 to it, because every sample since the last trigger gets an additional 20 timeunits latency.

In that sense, `self._current_regret` should keep track of the number of seen samples. It should be 10, since it does not matter what the regret value within that batch was. It should only matter that all 10 samples get another 20 timeunits regret when the next batch comes. But since new_regret is based on the timestamps, I don't see how this is possible. Might be easier to discuss this in person on the whiteboard.

Reply:
- `self._current_regret` reflects the most recent y value of the `number of samples since last trigger` curve. 
- `self._cumulative_latency_regret` reflects the area under this curve since the last trigger.

I hope this image helps
<img width="973" alt="image" src="https://github.com/user-attachments/assets/19947eac-bd5d-417f-a110-0e9deb616613">

Reply:
"self._current_regret reflects the most recent y value of the number of samples since last trigger curve."

That would be what I expect to see here. In my example, that would be 10 (because we already saw 10 samples, and each of them gets an additional regret of 20). However, in line 72, we do `self._current_regret += new_regret`. And new_regret is not equal to the number of samples, but instead a per-sample regret value.

Reply:
in the `data-incorporation-latency-cost-trigger` the regret is `1` for every sample, the time is only relevent for the `regret_latency`. I guess this definition of `regret` is a bit unintuitive as one might expect that regret already captures the time dimension. If you have a better name for the abstraction from data amount (or number of misclassifications in the second policy), I'm happy to adopt that.

Reply:
Hm, what would be the value of this in case of number of misclassifications? Because for these, we don't have a latency, right? It's just a linear curve? Hence, I would think in that case, regret would also be 1, and we should not have `old_regret_latency = self._current_regret * period_duration`. I am basically unsure in what other cases despite the latency that we already discussed this latency tracker will be used and what other values than 1 this can take and why

Reply:
I suspect that only using misclassification would yield the same problems as in the data incorporation latency case. Therefore, I was also planning to take this latency sum approach. For the base regret value, we I'd therfore use number of misclassifications and area under curve would be misclassification integration latency.
I'm not 100% sure yet how I will implement that on the trigger side. As we don't have access to what exact samples were misclassified (not planning to refactor that in the evaluation logic) I will either use the `add_latency(...)` function which only takes the aggregated regret for the last interval or I'll use `add_latencies` like before with a `regret` value of `expected_misclassification_rate` for every sample (instead of the `1` in the base case)

Reply:
Overall, I think the current interface gives most flexibility to subsequent policy implementations / adaptions.